politeiad/backendv2/tstorebe/plugins/ticketvote/activevotes.go

```diff
@@ -0,0 +1,309 @@
+	// cache when the ticket pool is 41k tickets and when using an off
+	// premise dcrdata instance with minimal latency. Any functions
+	// that rely of this cache should fallback to fetching the
+	// commitment addresses manually in the event the cache has not
+	// been fully populated yet or has experienced unforeseen errors
+	// during creation (ex. network errors). If the initial job fails
+	// to complete it will not be retried.
+	Addrs map[string]string // [ticket]address
+}
+
+// VoteDetails returns the vote details from the active votes cache for the
+// provided token. If the token does not correspond to an active vote then nil
+// is returned.
+func (a *activeVotes) VoteDetails(token []byte) *ticketvote.VoteDetails {
+	a.RLock()
+	defer a.RUnlock()
+
+	t := hex.EncodeToString(token)
+	av, ok := a.activeVotes[t]
+	if !ok {
+		return nil
+	}
+
+	// Return a copy of the vote details
+	eligible := make([]string, len(av.Details.EligibleTickets))
+	for i, v := range av.Details.EligibleTickets {
+		eligible[i] = v
+	}
+	options := make([]ticketvote.VoteOption, len(av.Details.Params.Options))
+	for i, v := range av.Details.Params.Options {
+		options[i] = v
+	}
+	return &ticketvote.VoteDetails{
+		Params: ticketvote.VoteParams{
+			Token:            av.Details.Params.Token,
+			Version:          av.Details.Params.Version,
+			Type:             av.Details.Params.Type,
+			Mask:             av.Details.Params.Mask,
+			Duration:         av.Details.Params.Duration,
+			QuorumPercentage: av.Details.Params.QuorumPercentage,
+			PassPercentage:   av.Details.Params.PassPercentage,
+			Options:          options,
+			Parent:           av.Details.Params.Parent,
+		},
+		PublicKey:        av.Details.PublicKey,
+		Signature:        av.Details.Signature,
+		StartBlockHeight: av.Details.StartBlockHeight,
+		StartBlockHash:   av.Details.StartBlockHash,
+		EndBlockHeight:   av.Details.EndBlockHeight,
+		EligibleTickets:  eligible,
+	}
+}
+
+// EligibleTickets returns the eligible tickets from the active votes cache for
+// the provided token. If the token does not correspond to an active vote then
+// nil is returned.
+func (a *activeVotes) EligibleTickets(token []byte) map[string]struct{} {
+	a.RLock()
+	defer a.RUnlock()
+
+	t := hex.EncodeToString(token)
+	av, ok := a.activeVotes[t]
+	if !ok {
+		return nil
+	}
+
+	// Return a map of the eligible tickets
+	eligible := make(map[string]struct{}, len(av.Details.EligibleTickets))
+	for _, v := range av.Details.EligibleTickets {
+		eligible[v] = struct{}{}
+	}
+
+	return eligible
+}
+
+// VoteIsDuplicate returns whether the vote has already been cast. This
+// function will panic if the provided token does not correspond to a record in
+// the active votes cache.
+func (a *activeVotes) VoteIsDuplicate(token, ticket string) bool {
+	a.RLock()
+	defer a.RUnlock()
+
+	av, ok := a.activeVotes[token]
+	if !ok {
+		// This should not happen
+		e := fmt.Errorf("active vote not found %v", token)
+		panic(e)
```

Comment:
Really? Is this terminal?

Also, move the Errorf into the panic if we keep it.

Reply:
tracked in https://github.com/decred/politeia/issues/1369